seismiqb/src/samplers.py

```diff
@@ -0,0 +1,347 @@
+            - 2 where only xline-oriented crops can be sampled.
+            - 3 where both types of crop orientations can be sampled.
+        """
+        flags = self.flags
+        i_matrix = np.full_like(self.horizon.full_matrix, 0.0)
+        i_points = self.locations[~flags]
+        i_matrix[i_points[:, 0], i_points[:, 1]] = 1
+
+        x_matrix = np.full_like(self.horizon.full_matrix, 0.0)
+        x_points = self.locations[flags]
+        x_matrix[x_points[:, 0], x_points[:, 1]] = 2
+
+        matrix = i_matrix + x_matrix
+        matrix[matrix == 0.0] = np.nan
+        return matrix
+
+
+
+class SeismicSampler(Sampler):
+    """ Mixture of samplers on multiple cubes with label-samplers.
+    Used to sample crop locations in (cube_id, i_start, x_start, h_start, i_stop, x_stop, h_stop) format, 
+    potentially with additional colums.
+
+    Parameters
+    ----------
+    labels : dict
+        Dictionary with nested lists that contains labels.
+    shape : tuple
+        Shape of crop locations to generate.
+    proportions : sequence, optional
+        Proportion of each cube in the resulting mixture.
+    baseclass : type
+        Class for label samplers.
+    threshold : float
+        Minimum proportion of labeled points in each sampled location.
+    filtering_matrix : np.ndarray, optional
+        Map of points to remove from potentially generated locations.
+    kwargs : dict
+        Other parameters of initializing label samplers.
+    """
+    def __init__(self, labels, shape, proportions=None, baseclass=HorizonSampler,
+                 threshold=0.05, filtering_matrix=None, **kwargs):
+
+        sampler = 0 & ConstantSampler(0, dim=baseclass.dim)
+        samplers = IndexedDict({idx: [] for idx in labels.keys()})
+        proportions = proportions or [1 / len(labels) for _ in labels]
+
+        for (idx, list_labels), p in zip(labels.items(), proportions):
+            cube_sampler = 0 & ConstantSampler(0, dim=baseclass.dim)
+            shape_ = shape[idx] if isinstance(shape, dict) else shape
+            threshold_ = threshold[idx] if isinstance(threshold, dict) else threshold
+            filtering_matrix_ = filtering_matrix[idx] if isinstance(filtering_matrix, dict) else filtering_matrix
+
+            for label in list_labels:
+                label_sampler = baseclass(label, shape=shape_, threshold=threshold_,
+                                          filtering_matrix=filtering_matrix_, **kwargs)
+                cube_sampler = cube_sampler | label_sampler
+                samplers[idx].append(label_sampler)
+
+            sampler = sampler | (p & cube_sampler)
+
+        self.sampler = sampler
+        self.samplers = samplers
+        self.proportions = proportions
+
+    def sample(self, size):
+        """ Generate exactly `size` locations. """
+        return self.sampler.sample(size)
+
+    def __call__(self, size):
+        return self.sampler.sample(size)
```

Comment:
```suggestion
        return self.sample(size)
```
or even `__call__ = sample `